modules/tensor_mechanics/src/actions/PrescribedBCAction.C

```diff
@@ -0,0 +1,190 @@
+  params.addRequiredParam<std::vector<BoundaryName> >("boundary", "Boundary on which the boundary condition is applied.");
+  return params;
+}
+
+PrescribedBCAction::PrescribedBCAction(const InputParameters & params) :
+  Action(params),
+  _data_file(getParam<FileName>("data_file"))
+{
+  _number = 1;
+}
+
+void
+PrescribedBCAction::act()
+{
+  Real gamma = getParam<Real>("gamma");
+  Real beta = getParam<Real>("beta");
+  bool prescribe_displacement = getParam<bool>("prescribe_displacement");
+
+  if (isParamValid("time_step") && prescribe_displacement)
+    mooseError("The time_step option will remain unused when displacement is prescribed. Please ensure the time step used in the data_file matches the dt and dtmin in the Executioner block.");
+
+  std::vector<Real> time;
+  std::vector<Real> data;
+  std::vector<Real> accel;
+
+  parseColumns(time, data);
+
+  Real dt = time[1] - time[0];
+
+  if (prescribe_displacement)
+  {
+    // Backward Euler time integration
+    std::vector<Real> Euler_vel(data.size(), 0.0);
+    accel.resize(data.size());
+
+    for (unsigned int i = 1; i < data.size(); ++i)
+    {
+      Euler_vel[i] = (data[i] - data[i-1]) / dt;
+      accel[i] = (Euler_vel[i] - Euler_vel[i-1]) / dt;
+    }
+  }
+  else
+  {
+    _number = time.size();
+
+    if (isParamValid("time_step"))
+    {
+      Real new_dt = getParam<Real>("time_step");
+      unsigned int N = dt / new_dt; // number of new points in each dt
+      if (std::abs(new_dt * N - dt) > 1e-8)
+        mooseError("An integer multiple of the 'time_step' (" << new_dt << ") should give the time step used in the data file (" << dt << ").");
+
+      accel.resize((_number - 1) * N + 1);
+      std::vector<Real> time_new((_number - 1) * N + 1);
+      for (unsigned int i = 0; i < _number - 1; ++i)
+      {
+        accel[i * N] = data[i];
+        time_new[i * N] = time[i];
+        for (unsigned int j = 0; j < N; ++j)
+        {
+          Real x = time[i] + j * new_dt;
+          time_new[i * N + j] = x;
+          accel[i * N + j] = LinearInterpolate(data[i], data[i+1], time[i], time[i+1], x);
+        }
+      }
+      accel[(_number-1) * N] = data[_number - 1];
+      time_new[(_number - 1) * N] = time[_number - 1];
+      time.resize(accel.size());
+      time = time_new;
+      dt = new_dt;
+    }
+    else
+      accel = data;
+  }
+
+  //  Newmark time integration
+  std::vector<Real> disp_final(accel.size(),0.0);
+  std::vector<Real> vel_final(accel.size(),0.0);
+
+  for (unsigned int i = 1; i < accel.size(); ++i)
+  {
+    vel_final[i] = vel_final[i-1] + (1-gamma) * dt * accel[i-1] + gamma * dt * accel[i];
+    disp_final[i] = disp_final[i-1] + dt * vel_final[i-1] + (0.5 - beta) * dt * dt * accel[i-1] + beta * dt * dt * accel[i];
+  }
+
+  InputParameters params = _factory.getValidParams("PiecewiseLinear");
+  params.set<std::vector<Real> > ("x") = time;
+  params.set<std::vector<Real> > ("y") = disp_final;
```

Comment:
no space before `(` (and since C++11 the space in `> >` is not necessary anymore either).